django-rgd/rgd/models/utils.py

```diff
@@ -0,0 +1,140 @@
+
+    Parameters
+    ----------
+    precompute_url_checksum : bool
+        Precompute the checksum of URL files to prevent duplicate files at different addresses.
+
+    """
+    parsed = urlparse(url)
+    if parsed.scheme not in ['https', 'http', 's3']:
+        raise ValidationError(f'Not a supported URL scheme ({parsed.scheme}) for URL: {url}')
+    if precompute_url_checksum:
+        checksum = compute_checksum_url(url)
+        kwargs.setdefault('checksum', checksum)
+    try:
+        file_entry = ChecksumFile.objects.get(url=url, collection=collection, **kwargs)
+        if file_entry.status != Status.SUCCEEDED and file_entry.status != Status.SKIPPED:
+            # If in a failed state, rerun tasks
+            file_entry.save()
+        return file_entry, False
+    except ChecksumFile.DoesNotExist:
+        defaults = defaults.copy()
+        defaults.update(kwargs)
+        file_entry = ChecksumFile(**defaults)
+        file_entry.url = url
+        file_entry.type = FileSourceType.URL
+        file_entry.collection = collection
+        file_entry.save()
+        return file_entry, True
+
+
+def get_or_create_checksumfile_file_field(
+    file_handle: io.BufferedIOBase, collection: Collection = None, defaults: dict = None, **kwargs
+):
+    """Get or create ChecksumFile from an open file handle and Collection."""
+    if 'checksum' not in kwargs:
+        file_handle.seek(0)
+        checksum = compute_hash(file_handle)
+        kwargs.setdefault('checksum', checksum)
+    file_entry, created = get_or_create_no_commit(
+        ChecksumFile, collection=collection, defaults=defaults, **kwargs
+    )
+    if created:
+        # Since we already computed the checksum, prevent tasks from recomputing
+        file_entry.skip_signal = True
+        file_handle.seek(0)
+        file_entry.file.save('test', file_handle)
+        file_entry.save()
+        file_entry.skip_signal = False
+    return file_entry, created
+
+
+def get_or_create_checksumfile(
+    collection: Collection = None, precompute_url_checksum=False, defaults: dict = None, **kwargs
+):
+    """Get or create a ChecksumFile.
+
+    Parameters
+    ----------
+    precompute_url_checksum : bool
+        Precompute the checksum of URL files to prevent duplicate files at different addresses.
+
+    Notes
+    -----
+    * URL files are only evaluated by their URL and Collection. Use the `url` keyword argument.
+    * FILE_FIELD files are evaluated by their checksum value and Collection. Use the `file` keyword argument.
+
+    Since files cannot belong to multiple Collections, we have to validate
+    against a Collection. If no Collection is given, then the Collection
+    of any existing duplicate file must be None. Otherwise, the file will be
+    created to prevent cross-collection permission issues.
+
+    """
+    if defaults is None:
+        defaults = {}
+    if 'collection' in kwargs or 'collection' in defaults:
```

Comment:
> In this scenario, having collection in defaults would make sense but we are going to ignore that for now...

I'm ignoring this, because I don't have the bandwidth to implement a full solution that would check the user's available collections for that file and only create the file if it existed in a collection that user had access to. This would require https://github.com/ResonantGeoData/ResonantGeoData/pull/529#discussion_r717014302 and would require hooking up to the middleware to see what user is making the request.

Overall, too many edge cases, and we can address later if needed

Reply:
I see, my main confusion was that the `collection` argument can be `None`, so if it was solely specified through `kwargs` or `defaults`, it would still throw this error. I'm much less familiar with any edge cases or uses of this function though, so I'll leave it at your discretion.

Reply:
Ah yeah, `collection=None` has a specific meaning here and we wouldn't want to pull from `defaults` or `kwargs` without addressing a bigger set of edge cases.